core/capabilities/syncer.go

```diff
@@ -56,26 +71,486 @@ var defaultStreamConfig = p2ptypes.StreamConfig{
 const maxRetryCount = 60
 
 // RegistrySyncer updates local Registry to match its onchain counterpart
-func NewRegistrySyncer(peerWrapper p2ptypes.PeerWrapper, registry core.CapabilitiesRegistry, dispatcher remotetypes.Dispatcher, lggr logger.Logger,
-	networkSetup HardcodedDonNetworkSetup) *registrySyncer {
+func NewRegistrySyncer(
+	peerWrapper p2ptypes.PeerWrapper,
+	registry core.CapabilitiesRegistry,
+	dispatcher remotetypes.Dispatcher,
+	lggr logger.Logger,
+	networkSetup HardcodedDonNetworkSetup,
+	relayer contractReaderFactory,
+	registryAddress string,
+) (*registrySyncer, error) {
+	stopCh := make(services.StopChan)
+	ctx, _ := stopCh.NewCtx()
+	reader, err := newRemoteRegistryReader(ctx, relayer, registryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	return newRegistrySyncer(
+		stopCh,
+		peerWrapper,
+		registry,
+		dispatcher,
+		lggr,
+		networkSetup,
+		reader,
+	), nil
+}
+
+func newRegistrySyncer(
+	stopCh services.StopChan,
+	peerWrapper p2ptypes.PeerWrapper,
+	registry core.CapabilitiesRegistry,
+	dispatcher remotetypes.Dispatcher,
+	lggr logger.Logger,
+	networkSetup HardcodedDonNetworkSetup,
+	reader reader,
+) *registrySyncer {
 	return &registrySyncer{
+		stopCh:       stopCh,
 		peerWrapper:  peerWrapper,
 		registry:     registry,
 		dispatcher:   dispatcher,
 		networkSetup: networkSetup,
 		lggr:         lggr,
+		reader:       reader,
 	}
 }
 
 func (s *registrySyncer) Start(ctx context.Context) error {
-	s.wg.Add(1)
-	go s.launch(context.Background())
+	s.wg.Add(2)
+	go s.launch()
+	go s.syncLoop()
+	return nil
+}
+
+func (s *registrySyncer) syncLoop() {
+	defer s.wg.Done()
+
+	ctx, cancel := s.stopCh.NewCtx()
+	defer cancel()
+
+	ticker := time.NewTicker(defaultTickInterval)
+	defer ticker.Stop()
+
+	// Sync for a first time outside the loop; this means we'll start a remote
+	// sync immediately once spinning up syncLoop, as by default a ticker will
+	// fire for the first time at T+N, where N is the interval.
+	s.lggr.Debug("starting initial sync with remote registry")
+	err := s.sync(ctx)
+	if err != nil {
+		s.lggr.Errorw("failed to sync with remote registry", "error", err)
+	}
+
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		case <-ticker.C:
+			s.lggr.Debug("starting regular sync with the remote registry")
+			err := s.sync(ctx)
+			if err != nil {
+				s.lggr.Errorw("failed to sync with remote registry", "error", err)
+			}
+		}
+	}
+}
+
+func (s *registrySyncer) sync(ctx context.Context) error {
```

Comment:
I think there should be a split between `sync`, which gets all the state and augments it with local versions, and launcher (Makram's name), which does `addRemoteCapabilities` and `enableExternalAccess`. This way, all state transformation would effectively be a pure function (which could potentially be merged with the reader), which does the "onchain state" => "augmented onchain state" transformation (would be easier to test, too). And then side-effects would happen in the launcher component.

Reply:
This will also future-proof for CCIP's use case.

Reply:
😓 I really don't want to consider such a big change at this stage... When I discussed the implementation of this with @bolekk we agreed to do the minimum to meet the Keystone use case.

Reply:
Right now the CCIP launcher is expected to have the full registry state & will do the required filtering / capability version matching and launch the appropriate process.

We basically just need a way to access `readerState` below

Reply:
> We basically just need a way to access readerState below

Reply:
> We basically just need a way to access readerState below

That's the proposal - splitting state reading and (internal) augmentation from the launching of services. I also don't think it needs to be a huge change, just moving some code around 🤷‍♂️

Reply:
Yup @DeividasK sounds good to me

Reply:
What do you understand by "internal augmentation"? Is this determining `myWorkflowDONs`/`remoteWorkflowDONs` etc? Or something else?

Reply:
> What do you understand by "internal augmentation"? Is this determining myWorkflowDONs/remoteWorkflowDONs etc?

Yes, this

Reply:
Its not elegant but can we add some mutex-protected getter for the latest registry state?

Reply:
Just to get this merged in & we can clean up later?